compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp

```diff
@@ -30,6 +32,80 @@ namespace mlir::iree_compiler {
 
 namespace {
 
+/// Folds IR resembling:
+/// ```
+///   %20 = vector.transfer_write %19, %16[%c0], %17 {in_bounds = [true]} 
+//      : vector<128xf16>, tensor<?xf16> 
+//    %21 = vector.transfer_read %20[%c0], %cst_2
+///     : tensor<?xf16>, vector<128xf16>
+/// ```
+/// into a simpler masked vector.transfer_read.
+/// After bufferization, this generally removes the need for materializing the
+/// write to memory.
+// TODO: Consider upstreaming
```

Comment:
Can be probably be upstreamed to TransferReadAfterWriteToBroadcast canonicalization pattern. The pattern currently just bails on inequal, but we could fix that.

Reply:
We discuss this point and there is a potential issue here due to pattern condition "convexity".
We want the `TransferReadAfterWriteToBroadcast` to only capture the broadcast case and bail otherwise.
Today the pattern also captures the non-broadcast case and is actually wrong in that case.
I'll send something to clean things up upstream but atm it seems we need a new pattern.
If we put everything in `TransferReadAfterWriteToBroadcast` then it becomes even harder to figure out how pattern sets intersect and who should have precedence.